node/src/chain/header_validation.rs

```diff
@@ -78,29 +80,50 @@ impl<'a, DB: database::DB> Validator<'a, DB> {
         candidate_block: &'a ledger::Header,
     ) -> anyhow::Result<()> {
         if candidate_block.version > 0 {
-            return Err(anyhow!("unsupported block version"));
+            anyhow::bail!("unsupported block version");
         }
 
         if candidate_block.hash == [0u8; 32] {
-            return Err(anyhow!("empty block hash"));
+            anyhow::bail!("empty block hash");
         }
 
         if candidate_block.height != self.prev_header.height + 1 {
-            return Err(anyhow!(
+            anyhow::bail!(
                 "invalid block height block_height: {:?}, curr_height: {:?}",
                 candidate_block.height,
                 self.prev_header.height,
-            ));
+            );
         }
 
         if candidate_block.prev_block_hash != self.prev_header.hash {
-            return Err(anyhow!("invalid previous block hash"));
+            anyhow::bail!("invalid previous block hash");
+        }
+
+        if candidate_block.timestamp > get_current_timestamp() {
+            anyhow::bail!("invalid future timestamp");
+        }
+
+        if candidate_block.timestamp < self.prev_header.timestamp {
+            anyhow::bail!("invalid timestamp");
+        }
+
+        if candidate_block.iteration < RELAX_ITERATION_THRESHOLD {
```

Comment:
If we have a candidate above `RELAX_ITERATION_THRESHOLD`, shouldn't we still have some `max_delta`?

Reply:
Relaxed iterations have no time limits, they move on to the next step only when the current step reaches a result. 
In the case of Proposal, the step waits indefinitely for a Candidate block.
So there shouldn't be any timestamp check beyond the condition of being subsequent to the previous block

Reply:
max_delta is depending to the iteration

Indeed is MAX_ITERATION_TIMEOUT*ITERATION